bundles/automation/org.eclipse.smarthome.automation.module.core/src/main/java/org/eclipse/smarthome/automation/module/core/provider/AnnotationActionModuleTypeHelper.java

```diff
@@ -0,0 +1,233 @@
+            }
+        }
+        return outputs;
+    }
+
+    public ActionType buildModuleType(String UID, Map<String, Set<ModuleInformation>> moduleInformation) {
+        Set<ModuleInformation> mis = moduleInformation.get(UID);
+        List<ConfigDescriptionParameter> configDescriptions = new ArrayList<>();
+
+        if (mis != null && mis.size() > 0) {
+            ModuleInformation mi = (ModuleInformation) mis.toArray()[0];
+
+            ActionModuleKind kind = ActionModuleKind.SINGLE;
+            if (mi.getConfigName() != null && mi.getThingUID() != null) {
+                logger.error(
+                        "ModuleType with UID {} has thingUID ({}) and multi-service ({}) property set, ignoring it.",
+                        UID, mi.getConfigName(), mi.getThingUID());
+                return null;
+            } else if (mi.getConfigName() != null) {
+                kind = ActionModuleKind.SERVICE;
+            } else if (mi.getThingUID() != null) {
+                kind = ActionModuleKind.THING;
+            }
+
+            ConfigDescriptionParameter configParam = buildConfigParam(mis, kind);
+            if (configParam != null) {
+                configDescriptions.add(configParam);
+            }
+
+            ActionType at = new ActionType(UID, configDescriptions, mi.getLabel(), mi.getDescription(), mi.getTags(),
+                    mi.getVisibility(), mi.getInputs(), mi.getOutputs());
+            return at;
+        }
+        return null;
+    }
+
+    private ConfigDescriptionParameter buildConfigParam(Set<ModuleInformation> moduleInformations,
+            ActionModuleKind kind) {
+        List<ParameterOption> options = new ArrayList<>();
+        if (kind == ActionModuleKind.SINGLE) {
+            if (moduleInformations.size() == 1) {
+                if (((ModuleInformation) moduleInformations.toArray()[0]).getConfigName() == null
+                        && ((ModuleInformation) moduleInformations.toArray()[0]).getThingUID() == null) {
+                    // we have a single service and thus no configuration at all
+                    return null;
+                }
+            }
+        } else if (kind == ActionModuleKind.SERVICE) {
+            for (ModuleInformation mi : moduleInformations) {
+                String configName = mi.getConfigName();
+                options.add(new ParameterOption(configName, null));
+            }
+            return ConfigDescriptionParameterBuilder.create(CONFIG_PARAM, Type.TEXT).withLabel(SELECT_SERVICE_LABEL)
+                    .withOptions(options).withLimitToOptions(true).withRequired(true).build();
+        } else if (kind == ActionModuleKind.THING) {
+            for (ModuleInformation mi : moduleInformations) {
+                String thingUID = mi.getThingUID();
+                options.add(new ParameterOption(thingUID, null));
+            }
+            return ConfigDescriptionParameterBuilder.create(CONFIG_PARAM, Type.TEXT).withLabel("Select Thing")
```

Comment:
If only we would have a good documentation on how to localize ANY string... int he documentation I only stumbled across localizing thing types...

Reply:
Please provide a `@text/<yourKey>` placeholder here and a translation file in `ESH-INF/i18n`. If I'm not mistaken the `ModuleTypeResourceBundleProvider` should take care of translating this keys.

Reply:
Annotated ModuleTypes can now be translated by providing `@text/id` expressions inside of the annotations. These texts however, are tricky to translate because its unclear where to provide the translation files and at this point of the code we do not know the locale into which we should translate. Thus, I am leaving this out of the PR.